From 88a4e503c4d7093859c866f909136bc3be2257d1 Mon Sep 17 00:00:00 2001
From: Bhagavathi Perumal S <quic_bperumal@quicinc.com>
Date: Sat, 17 Jun 2023 12:23:15 +0530
Subject: [PATCH] hostapd: use after free fix

fix the pointer usage after memory free on the wpa_bss_update function

Signed-off-by: Bhagavathi Perumal S <quic_bperumal@quicinc.com>
---
 wpa_supplicant/bss.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/wpa_supplicant/bss.c b/wpa_supplicant/bss.c
index a3da86c..a0ec7d6 100644
--- a/wpa_supplicant/bss.c
+++ b/wpa_supplicant/bss.c
@@ -725,8 +725,8 @@ wpa_bss_update(struct wpa_supplicant *wpa_s, struct wpa_bss *bss,
 		struct wpa_bss *nbss;
 		struct dl_list *prev = bss->list_id.prev;
 		dl_list_del(&bss->list_id);
-		nbss = os_realloc(bss, sizeof(*bss) + res->ie_len +
-				  res->beacon_ie_len);
+		nbss = os_malloc(sizeof(*bss) + res->ie_len +
+				 res->beacon_ie_len);
 		if (nbss) {
 			unsigned int i;
 			for (i = 0; i < wpa_s->last_scan_res_used; i++) {
@@ -738,6 +738,7 @@ wpa_bss_update(struct wpa_supplicant *wpa_s, struct wpa_bss *bss,
 			if (wpa_s->current_bss == bss)
 				wpa_s->current_bss = nbss;
 			wpa_bss_update_pending_connect(wpa_s, bss, nbss);
+			os_free(bss);
 			bss = nbss;
 			os_memcpy(bss->ies, res + 1,
 				  res->ie_len + res->beacon_ie_len);
-- 
2.17.1

